tools/nimgrep.nim

```diff
@@ -223,242 +324,357 @@ proc writeArrow(s: string) =
 proc blockHeader(filename: string, line: int|string, replMode=false) =
   if replMode:
     writeArrow("     ->\n")
-  elif newLine:
+  elif newLine and optFilenames notin options:
     if oneline:
       printBlockFile(filename)
       printBlockLineN(":" & $line & ":")
     else:
       printBlockLineN($line.`$`.align(alignment) & ":")
     stdout.write("\n")
 
-proc lineHeader(filename: string, line: int|string, isMatch: bool) =
+proc newLn(curCol: var Column) =
+  stdout.write("\n")
+  curCol.file = 0
+  curcol.terminal = 0
+
+proc lineHeader(filename: string, line: int|string, isMatch: bool,
+                curCol: var Column) =
   let lineSym =
     if isMatch: $line & ":"
     else: $line & " "
-  if not newLine:
+  if not newLine and optFilenames notin options:
     if oneline:
       printFile(filename)
       printLineN(":" & lineSym, isMatch)
+      curcol.terminal += filename.len + 1 + lineSym.len
     else:
       printLineN(lineSym.align(alignment+1), isMatch)
-    stdout.write(" ")
+      curcol.terminal += lineSym.align(alignment+1).len
+    stdout.write(" "); curCol.terminal += 1
 
-proc printMatch(fileName: string, mi: MatchInfo) =
-  let lines = mi.match.splitLines()
-  for i, l in lines:
+proc printMatch(fileName: string, mi: MatchInfo, curCol: var Column) =
+  let sLines = mi.match.splitLines()
+  for i, l in sLines:
     if i > 0:
-      lineHeader(filename, mi.lineBeg + i, isMatch = true)
-    writeColored(l)
-    if i < lines.len - 1:
-      stdout.write("\n")
+      lineHeader(filename, mi.lineBeg + i, isMatch = true, curCol)
+    if curCol.terminal < limitChar:
+      writeColored(l)
+    else:
+      curCol.overflowMatches += 1
+    if i < sLines.len - 1:
+      newLn(curCol)
+  curCol.terminal += mi.match.len
+  curCol.file += mi.match.len
+
+const matchPaddingFromRight = 10
+let ellipsis = "..."
+
+proc reserveChars(mi: MatchInfo): int =
+  if optLimitChars in options or optFit in options:
+    let patternChars = afterPattern(mi.match, 0) + 1
+    result = patternChars + ellipsis.len + matchPaddingFromRight
+  else:
+    result = 0
+
+proc printRaw(c: char, curCol: var Column, allowTabs = true) =
+  # print taking into account tabs and optOnlyAscii
+  if c == '\t':
+    if allowTabs:
+      let spaces = 8 - (curCol.file mod 8)
+      curCol.file += spaces
+      curCol.terminal += spaces
+      if optOnlyAscii:
+        printSpecial " "
+        stdout.write " ".repeat(spaces-1)
```

Comment:
`stdout.write` (which results in a syscall) should be much slower than using your own line buffer (a single `buffer: string` that you reuse and reset) that you flush after each new line

Reply:
I believe it's OK, it does not result in a syscall. write(File, string) in io.nim:244 call writeBuffer on :199, which calls fwrite. fwrite from C is already buffered. To flush one needs to call stdout.flushFile

Reply:
fwrite still incurs overhead IIRC; could this explain (parts of) the 25x slowdown in https://github.com/nim-lang/Nim/pull/15612#discussion_r509846344 ?

Reply:
EDIT: indeed, see the stackoverflow answer I just posted in https://stackoverflow.com/a/64492448/1426932 which demonstrates this

Reply:
I've run your test on my Linux Debian 10 and I see completely different results:
case 1 — 3.4 s
case 2 — 5.4 s
So using in-app buffering actually decreased speed of the program significantly!
My first guess is that fwrite is implemented inefficiently in OSX or may be buffer size is too small.

Reply:
In updated version version in-app buffering is implemented (though mainly for different purpose), it may increase the speed in your case.